File: frontend/projects/irrtirer/src/app/shared/mosaic-generator/orderer/section-layout/possible-layout.ts
import { Tile } from "../../models/tile";
import { TileTransform } from "../../models/tile-transform.model";
import { SectionModel } from "../../sectors/section.model";
import { LayoutTilesRate } from "./layout-tiles-rate.model";

export class PossibleLayout {   
    public readonly section: SectionModel;

    public rate: number = Number.MIN_VALUE;

    public readonly layoutTilesRates: LayoutTilesRate = new LayoutTilesRate();

    public usedTilesIndexes: Set<number>;
    public tilesAvalibleForLayout: Tile[];

    constructor(section: SectionModel) {
        this.section = section;
    }

    public getTilesLayout(): TileTransform[] {
        return this.layoutTilesRates.getKeys();
    }
}

File: backend/Irrtirer/Irrtirer.Generator/Orderer/SectionLayout/PossibleLayout.cs
using Irrtirer.Generator.MosaicStructure;
using Irrtirer.Generator.Tiles;
using System;
using System.Collections.Generic;
using System.Linq;

namespace Irrtirer.Generator.Orderer.SectionLayout
{
    internal class PossibleLayout : ICloneable
    {
        public SectionModel Section { get; private set; }

        public float Rate { get; set; } = float.MinValue;

        public LayoutTilesRate LayoutTilesRates { get; private set; } = new LayoutTilesRate();
        public List<TileTransform> TilesLayout { get { return LayoutTilesRates.Keys.ToList(); } }

        public HashSet<int> UsedTilesIndexes { get; set; }
        public Tile[] TilesAvalibleForLayout { get; set; }

        private PossibleLayout() { }

        public PossibleLayout(SectionModel section)
        {
            Section = section;
        }

        public object Clone()
        {
            return new PossibleLayout()
            {
                Section = Section,
                Rate = Rate,
                LayoutTilesRates = (LayoutTilesRate)LayoutTilesRates.Clone(),
                UsedTilesIndexes = UsedTilesIndexes.ToHashSet(),
                TilesAvalibleForLayout = TilesAvalibleForLayout
            };
        }
    }
}
